### PR TITLE
Implement the second wave of per-reconciler leaderelection

### PR DESCRIPTION
**Pull request by mattmoor**
Detailed design: https://docs.google.com/document/d/1i_QHjQO2T3SNv49xjZLWlivcc0UvZN1Tbw2NKxThkyM/edit

Issue: https://github.com/knative/pkg/issues/1181

WIP until #1301 lands

### CODE REVIEW

File: webhook/certificates/certificates.go [hunk 1]
```diff
@@ -37,17 +39,27 @@ const (
...
+var _ pkgreconciler.LeaderAware = (*reconciler)(nil)
 
 // Reconcile implements controller.Reconciler
 func (r *reconciler) Reconcile(ctx context.Context, key string) error {
-	return r.reconcileCertificate(ctx)
+	if r.IsLeaderFor(types.NamespacedName{
+		Namespace: system.Namespace(),
+		Name:      r.secretName,
+	}) {
+		// only reconciler the certificate when we are leader.
```
Comment (vagababov) [inline webhook/certificates/certificates.go:59]: looks like this is immutable? May be make it a member? Or perhaps it should use `key` instead?

File: webhook/configmaps/configmaps.go [hunk 2]
```diff
@@ -58,13 +61,19 @@ type reconciler struct {
...
 var _ controller.Reconciler = (*reconciler)(nil)
+var _ pkgreconciler.LeaderAware = (*reconciler)(nil)
 var _ webhook.AdmissionController = (*reconciler)(nil)
 var _ webhook.StatelessAdmissionController = (*reconciler)(nil)
 
 // Reconcile implements controller.Reconciler
 func (ac *reconciler) Reconcile(ctx context.Context, key string) error {
 	logger := logging.FromContext(ctx)
 
+	if !ac.IsLeaderFor(types.NamespacedName{Name: ac.name}) {
```
Comment (vagababov) [inline webhook/configmaps/configmaps.go:72]: Namespace?

File: webhook/configmaps/configmaps.go [hunk 3]
```diff
@@ -58,13 +61,19 @@ type reconciler struct {
...
+		logger.Debugf("Skipping key %q, not the leader.", key)
```
Comment (vagababov) [inline webhook/configmaps/configmaps.go:73]: Key is not part of the leader selection per code.

File: webhook/certificates/certificates.go (hunk 1 above)
Comment (mattmoor) [reply webhook/certificates/certificates.go]: I moved `key` to be a member of all of the singleton reconcilers.

File: webhook/configmaps/configmaps.go (hunk 2 above)
Comment (mattmoor) [reply webhook/configmaps/configmaps.go]: This is reconciling the validating webhook, so no namespace.

File: webhook/configmaps/configmaps.go (hunk 3 above)
Comment (mattmoor) [reply webhook/configmaps/configmaps.go]: All of these should now print `ac.key`

File: webhook/certificates/controller.go [hunk 4]
```diff
@@ -68,7 +70,7 @@ func NewController(
 
 	// Reconcile when the cert bundle changes.
 	secretInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
-		FilterFunc: controller.FilterWithNameAndNamespace(system.Namespace(), wh.secretName),
+		FilterFunc: controller.FilterWithNameAndNamespace(key.Namespace, key.Name),
```
Comment (dprotaso) [inline webhook/certificates/controller.go:73]: lol the func name `FilterWithNameAndNamespace` doesn't match the args order

Comment (mattmoor) [reply webhook/certificates/controller.go:73]: Yeah, and it'd be nice if it took a `types.NamespacedName` 🙄

File: webhook/configmaps/configmaps.go (hunk 2 above)
Comment (vagababov) [reply webhook/configmaps/configmaps.go]: we're really abusing the type, whose comment says `mandatory namespace` 😁

**Pull request merged**